turbine/src/broadcast_stage/standard_broadcast_run.rs

```diff
@@ -148,20 +138,33 @@ impl StandardBroadcastRun {
                 );
         process_stats.num_merkle_data_shreds += data_shreds.len();
         process_stats.num_merkle_coding_shreds += coding_shreds.len();
+
         if let Some(shred) = data_shreds.iter().max_by_key(|shred| shred.index()) {
             self.chained_merkle_root = shred.merkle_root().unwrap();
             self.next_shred_index = shred.index() + 1;
         };
-        if self.next_shred_index > max_data_shreds_per_slot {
-            return Err(BroadcastError::TooManyShreds);
-        }
         if let Some(index) = coding_shreds.iter().map(Shred::index).max() {
             self.next_code_index = index + 1;
         };
-        if self.next_code_index > max_code_shreds_per_slot {
-            return Err(BroadcastError::TooManyShreds);
-        }
-        Ok((data_shreds, coding_shreds))
```

Comment:
I believe this was intentional in order to have the `entries_to_shreds_max` test coverage.
You can sort of replace that test with a `#[should_panic]` test but that is ugly and unnecessary.

In general it is more reasonable to just return error up the call stack and let the caller decide if the error is irrecoverable or a simple `error!` log or a metric suffices. The call site always has more context to choose which is the best.

Also putting panics deep in the call stack always has the risk that someone might add another call to those methods without realizing that the function might panic. Whereas an `Err` is explicit and forces the caller to either handle that error gracefully or abort if that is the right thing to do.

Reply:
I agree that, in general, the caller might have more information to handle an error better and/or provide additional context in the produced error message.
But this is not the case here.

`entries_to_shreds()` has a very strong connection with `process_receive_results()`.
You actually moved the code between the two in your latest change, and, to me, it is not entirely clear what is the exact responsibility of each function.

`process_receive_results()` is the single caller, and it is just calling `unwrap()` on the result.  So we already have the `panic!` call.  It is just hidden, and it only _seems_ that the error is handled.

The error messages I've added actually provide more information than before.

As for the test coverage, my impression is that it is not that useful.
`Shredder` is already checking the shred indices, and this test is covering a case that does not seem to be particularly important.
There is already a [panic in the `Shredder::entries_to_shreds()` call](https://github.com/anza-xyz/agave/blob/8c5a33a81a0504fd25d0465bed35d153ff84819f/ledger/src/shredder.rs#L110-L117).
The only case this function was checking is if we reach the very last shred, and we are recording an out of range values for `next_{code,shred}_index`.
And if we do that, `Shredder::entries_to_shreds()` will still fail, should it actually be invoked with those indices.

As for the moving `unwrap()` from parent into `entries_to_shreds()`, I think, the code is trivially correct.
A unit test is not really making it that much better.

You can argue that the test protects from a breakage in case the code is changed.
But considering how intertwined `entries_to_shreds()` and `process_receive_results()` are, I think it is quite unlikely.

We do have a modified version that calls `entries_to_shreds()` from multiple places.
And the only outcome is just more `unwrap()` calls.
So in the only actual case (well 3 cases) where `entries_to_shreds()` is reused, this error handling is not useful.

Reply:
I still don't see what necessitates this change. If a new caller wants to either ignore or panic on the error, both are already possible with current code. There is no reason to "force" a panic.

As a general rule, we should avoid panics deep in the call stacks and let the caller decide if the error is irrecoverable or not. It does not matter if there is only one caller right now because (a) the code always changes and (b) it is still more explicit to propagate the error up the call stack and let the (single) caller deal with it.

Reply:
> I still don't see what necessitates this change. If a new caller wants to either ignore or panic on the error, both are already possible with current code. There is no reason to "force" a panic.

The change makes the function API simpler.
As we are making other changes to this function in our fork, a simpler base simplifies our changes as well.

> As a general rule, we should avoid panics deep in the call stacks and instead let the caller decide if the error is irrecoverable or not. It does not matter if there is only one caller right now because (a) the code always changes and (b) it is still more explicit to propagate the error up the call stack and let the (single) caller deal with it.

My understanding of a "general rule" is that while it makes sense in a lot of cases, it is not always the case.
In this particular case, a panic in `process_receive_results()` is no better than a panic in `entries_to_shreds()`.

`entries_to_shreds()` is an internal function that is part of the implementation of `process_receive_results()`.  An `unwrap()` in `process_receive_results()` is, effectively, moving the panic just happens further away from the spot where the error actually happened.  Does not seem like an advantage to me.

Reply:
The code always changes and explicitly returning an error is safer in the context of those changes.

I am still not convinced this is good change. If it make things easier in the fork, one easy option is just to make this change in the fork as well.

Reply:
I spend time rebasing changes in the fork every week.
The more changes in the fork we have, the more of my time is spent managing them.
This is the reason I want to upstream changes in the first place.

Reply:
I am happy to merge changes which simplify rebasing on your side as long as we are not making a negative trade-off with respect to code quality or design.
imo this change in particular does not meet that requirement.